src/idnlookup.1
.\"
.\" Copyright (c) 2005 Fredrik Tolf <fredrik@dolda2000.com>
.\"
.\" This is free documentation; you can redistribute it and/or
.\" modify it under the terms of the GNU General Public License as
.\" published by the Free Software Foundation; either version 2 of
.\" the License, or (at your option) any later version.
.\"
.\" The GNU General Public License's references to "object code"
.\" and "executables" are to be interpreted as the output of any
.\" document formatting or typesetting system, including
.\" intermediate and printed output.
.\"
.\" This manual is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public
.\" License along with this manual; if not, write to the Free
.\" Software Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111,
.\" USA.
.\"
.TH IDNLOOKUP 1 "12 Jan 2006" "icmpdn 0.3" "ICMP Hostname Manual"
.SH NAME
idnlookup \- ICMP domain name resolver
.SH SYNOPSIS
.B idnlookup
[ \fB-ahT\fP ] [ \fB-t\fP \fItimeout\fP ] \fIhostname\fP
.SH DESCRIPTION
\fBidnlookup\fP is a utility to ask an Internet host for its fully
qualified domain name over ICMP. If successful, it will print the
host's domain name to standard output.
.P
\fIhostname\fP is the address of the Internet host to query. It can be
either a symbolic hostname, an IPv4 address in decimal notation, or an
IPv6 address in hexadecimal notation. If a symbolic hostname is
specified, all its addresses are looked up and tried in order.
.P
The remote host must be running software which can respond to the
requests for hostname. To the knowledge of the author, current
BSD-based systems can do this in the kernel. Linux systems must, as of
this writing, be running \fBicmpdnd\fP(8).
.SH EXIT STATUS
\fBidnlookup\fP will return 0 if the specified host was succesfully
queried. If illegal arguments were given, or if the host could, for
any reason, not be queried, or if another error occurred, it will
return 1.
.SH OPTIONS
.TP
.B -a
Normally, \fBidnlookup\fP will only print the first name that the host
returns. This option will cause all names to be printed, one per
line.
.TP
.B -h
Print a brief usage message and exit.
.TP
.B -T
Write the time-to-live in seconds that the remote host suggests on the
first line, before any names.
.TP
.B -t
Print an error message and exit if the remote host does not reply in
\fItimeout\fP milliseconds. The default is 3 seconds. Note that this
timeout will apply to all addresses, in case a symbolic hostname
resolves to more than one address.
.SH SECURITY
Since \fBidnlookup\fP uses ICMP, it needs to run as root. Normally, it
should be installed set-uid root in order for any user to be able to
ask hosts for their names. Although the author believes
\fBidnlookup\fP to be safe to install as set-uid root, he makes no
guarantees.
.SH BUGS
\fBidnlookup\fP does not yet implement domain name decompression, and
as such will scream and die at the sight of such names. This is
because the author is not sure from which point in the ICMP datagrams
the offsets specified in compressed domain names is to be counted. If
you have information on this, please mail the author. Until then, this
bug will hopefully not pose any large problems, since probably only
very few hosts will return more than one name, and then there is
little chance to compress it.
.P
Some GNU/Linux distributions still ship with a GNU C library
implementation that has a bug in the implementation of
\fBgetaddrinfo\fP(3), which may cause addresses resolved from a
symbolic hostname to be returned in the wrong order (i.e. not
consistent with RFC 3484). This will in turn cause \fBidnlookup\fP to
try addresses in the wrong order, if the hostname in question triggers
the \fBgetaddrinfo\fP(3) bug. This will normally not pose a problem,
but not too uncommonly, this will cause IPv4 addresses to be sorted
before IPv6 addresses. If the remote host listens for ICMP hostname
requests only on IPv6 (or if the Internet service provider for the
remote host blocks ICMP packets, which is wrong, but unfortunately not
all too uncommon), \fBidnlookup\fP will have to time out on the IPv4
address before being able to query the IPv6 address. This bug has been
fixed since, but as of this writing, only a few systems run the fixed
version of glibc.
.SH NOTES
The ICMP hostname tools also ship with a nameswitch module for ICMP
hostname lookups, which can be used to automatically look up hostnames
over ICMP when a program calls \fBgethostbyaddr\fP(3) or any related
function. See the \fBnss-icmp\fP(7) man page for more information.
.SH AUTHOR
Fredrik Tolf <fredrik@dolda2000.com>
.SH SEE ALSO
\fBicmpdnd\fP(8), \fBnss-icmp\fP(7), \fIRFC 1788\fP


ChangeLog
2006-01-15  Fredrik Tolf  <fredrik@dolda2000.com>
	* configure.ac: Bump version to 0.4
	* admin/icmpdnd: Fixed chkconfig values to ensure that icmpdnd is
	started after ypbind.

2006-01-13  Fredrik Tolf  <fredrik@dolda2000.com>
	* src/*.[1-9]: Added man pages

2005-06-27  Fredrik Tolf  <fredrik@dolda2000.com>
	* src/icmpdnd.c: Made it possible to set the name explicitly.
	* configure.ac: Advance version to 0.3

2005-05-25  Fredrik Tolf  <fredrik@dolda2000.com>

	* admin/Makefile.am: Don't install the RedHat rc script on all systems.
	* configure.ac: Change version to 0.2


src/idnlookup.c
/*
 *  idnlookup - ICMP Domain Name lookup utility for Linux
 *  Should be installed SUID root, even though I don't know if it's secure yet. :-)
 *  Copyright (C) 2005 Fredrik Tolf <fredrik@dolda2000.com>
 *  
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *  
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *  
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
*/
#include <stdlib.h>
#include <unistd.h>
#include <stdio.h>
#include <string.h>
#include <errno.h>
#include <netdb.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <netinet/ip.h>
#include <netinet/ip6.h>
#include <arpa/inet.h>
#include <sys/types.h>
#include <sys/poll.h>
#include <sys/time.h>
#ifdef HAVE_CONFIG_H
#include "config.h"
#endif

struct icmphdr {
    u_int8_t type;
    u_int8_t code;
    u_int16_t checksum;
};

struct reqhdr {
    u_int8_t type;
    u_int8_t code;
    u_int16_t checksum;
    u_int16_t id;
    u_int16_t seq;
};

struct rephdr {
    u_int8_t type;
    u_int8_t code;
    u_int16_t checksum;
    u_int16_t id;
    u_int16_t seq;
    int32_t ttl;
    
};

#define ICMP_NAMEREQ 37
#define ICMP_NAMEREP 38

unsigned char buf[65536];

/* DN decompression not yet implemented, since I don't know where to
 * begin counting the offset from -- the beginning of the ICMP
 * payload, or from the beginning of the DN data buffer? */
void printdn(FILE *f, unsigned char *dnbuf, size_t size, int onlyfirst)
{
    unsigned char *p;
    
    p = dnbuf;
    while(p - dnbuf < size) {
	while(*p != 0) {
	    if(*p & 0xc0) {
		fprintf(stderr, "domain name decompression not implemented, aborting\n");
		exit(1);
	    }
	    if((int)*p + (p - dnbuf) >= size) {
		fprintf(stderr, "malformed domain name\n");
		return;
	    }
	    fprintf(f, "%.*s", (int)*p, p + 1);
	    p += 1 + (int)*p;
	    if(*p != 0)
		fprintf(f, ".");
	}
	p++;
	fprintf(f, "\n");
	if(onlyfirst)
	    break;
    }
}

void cksum(void *hdr, size_t len)
{
    struct icmphdr *ih;
    u_int8_t *cb;
    int i;
    int b1, b2;
    
    ih = (struct icmphdr *)hdr;
    cb = (u_int8_t *)hdr;
    ih->checksum = 0;
    b1 = b2 = 0;
    for(i = 0; i < (len & ~1); i += 2) {
	b1 += cb[i];
	b2 += cb[i + 1];
    }
    if(i & 1)
	b1 += cb[len - 1];
    while(1) {
	if(b1 >= 256) {
	    b2 += b1 >> 8;
	    b1 &= 0xff;
	    continue;
	}
	if(b2 >= 256) {
	    b1 += b2 >> 8;
	    b2 &= 0xff;
	    continue;
	}
	break;
    }
    cb = (u_int8_t *)&ih->checksum;
    cb[0] = ~(u_int8_t)b1;
    cb[1] = ~(u_int8_t)b2;
}

void usage(void)
{
    fprintf(stderr, "usage: idnlookup [-hTa] [-t timeout] host\n");
}

int main(int argc, char **argv)
{
    int ret;
    int s, c;
    int id;
    int namelen;
    struct reqhdr req;
    struct rephdr rep;
    struct iphdr iphdr;
    size_t hdrlen;
    struct addrinfo *ai, *cai, aihint;
    struct pollfd pfd;
    struct timeval tvb, tvc;
    struct sockaddr_storage name;
    int timeout, dispttl, onlyfirst;
    int elapsed, timedout, found;
    
    timeout = 3000;
    dispttl = 0;
    onlyfirst = 1;
    while((c = getopt(argc, argv, "haTt:")) != -1) {
	switch(c) {
	case 't':
	    timeout = atoi(optarg);
	    break;
	case 'a':
	    onlyfirst = 0;
	    break;
	case 'T':
	    dispttl = 1;
	    break;
	case 'h':
	case '?':
	case ':':
	default:
	    usage();
	    exit((c == 'h')?0:1);
	}
    }
    
    if(argc - optind < 1) {
	usage();
	exit(1);
    }
    
    memset(&aihint, 0, sizeof(aihint));
    aihint.ai_socktype = SOCK_RAW;
    aihint.ai_protocol = IPPROTO_ICMP;
    ret = getaddrinfo(argv[optind], NULL, &aihint, &ai);
    
    for(cai = ai; cai != NULL; cai = cai->ai_next) {
	if((s = socket(cai->ai_family, SOCK_RAW, IPPROTO_ICMP)) < 0) {
	    perror("could not create raw socket");
	    exit(1);
	}
	
	id = random() % 65536;
	memset(&req, 0, sizeof(req));
	req.type = ICMP_NAMEREQ;
	req.id = htons(id);
	cksum(&req, sizeof(req));
	
	ret = sendto(s, &req, sizeof(req), 0, cai->ai_addr, cai->ai_addrlen);
	if(ret < 0) {
	    perror("sendto");
	    exit(1);
	} else if(ret != sizeof(req)) {
	    fprintf(stderr, "socket would not send entire packet\n");
	    exit(1);
	}
	
	timedout = 0;
	found = 0;
	gettimeofday(&tvb, NULL);
	while(1) {
	    pfd.fd = s;
	    pfd.events = POLLIN;
	    gettimeofday(&tvc, NULL);
	    elapsed = ((tvc.tv_sec - tvb.tv_sec) * 1000) + ((tvc.tv_usec - tvb.tv_usec) / 1000);
	    if(elapsed >= timeout) {
		timedout = 1;
		break;
	    }
	    ret = poll(&pfd, 1, timeout - elapsed);
	    if(ret < 0) {
		perror("idnlookup: reading data");
		exit(1);
	    }
	    
	    if(pfd.revents & POLLIN) {
		namelen = sizeof(name);
		ret = recvfrom(s, buf, sizeof(buf), 0, (struct sockaddr *)&name, &namelen);
		if(ret < 0) {
		    perror("idnlookup: receiving data");
		    exit(1);
		}
		
		if(name.ss_family != cai->ai_addr->sa_family)
		    continue;
		if(name.ss_family == AF_INET) {
		    if(memcmp(&(((struct sockaddr_in *)&name)->sin_addr), &(((struct sockaddr_in *)cai->ai_addr)->sin_addr), sizeof(struct in_addr)))
			continue;
		    if(ret < sizeof(iphdr) + sizeof(rep))
			continue;
		    hdrlen = sizeof(iphdr);
		    memcpy(&iphdr, buf, sizeof(iphdr));
		    if(iphdr.protocol != IPPROTO_ICMP)
			continue;
		} else if(name.ss_family == AF_INET6) {
		    if(memcmp(&(((struct sockaddr_in6 *)&name)->sin6_addr), &(((struct sockaddr_in6 *)cai->ai_addr)->sin6_addr), sizeof(struct in6_addr)))
			continue;
		    if(ret < sizeof(rep))
			continue;
		    hdrlen = 0;
		} else {
		    continue;
		}
		
		memcpy(&rep, buf + hdrlen, sizeof(rep));
		if(rep.type != ICMP_NAMEREP)
		    continue;
		if((ntohs(rep.id) != id) || (ntohs(rep.seq != 0)))
		    continue;
		
		found = 1;
		break;
	    }
	}
	
	close(s);
	
	if(found) {
	    if(dispttl)
		printf("%i\n", ntohl(rep.ttl));
	    printdn(stdout, buf + hdrlen + sizeof(rep), ret - hdrlen - sizeof(rep), onlyfirst);
	    break;
	}
    }
    
    if(timedout) {
	fprintf(stderr, "idnlookup: timeout\n");
	exit(1);
    }
    
    return(0);
}

/*
 * Local Variables:
 * compile-command: "make CFLAGS='-Wall -g'"
 * End:
 */
